spock-core/src/main/java/org/spockframework/compiler/WhereBlockRewriter.java

```diff
@@ -171,6 +263,30 @@ private void rewriteBinaryWhereStat(ListIterator<Statement> stats) throws Invali
         // neither of the other two
         throw notAParameterization(stat);
       }
+      // if derived data variable is multiplicand => error
+      if (COMBINED.equals(stat.getStatementLabel())) {
+        throw derivedDataVariablesCannotBeCombined(stat);
+      }
+      // if derived data variable is multiplier => error
+      int i = 0;
+      try {
+        while (stats.hasNext()) {
+          Statement next = stats.next();
+          i++;
+          if (COMBINED.equals(next.getStatementLabel())) {
+            throw derivedDataVariablesCannotBeCombined(stat);
+          }
+          // if not data table separator, we are done
+          // otherwise check next statement for combined label
+          if (!isDataTableSeparator(next)) {
+            break;
+          }
+        }
+      } finally {
+        for (int j = 0; j < i; j++) {
+          stats.previous();
+        }
+      }
```

Comment:
Extract a method.
Also maybe two because the code and comment is duplicated 
```java
// if derived data variable is multiplicand => error
if (COMBINED.equals(stat.getStatementLabel())) {
  throw derivedDataVariablesCannotBeCombined(stat);
}
```